src/Compiler/ES/Imperative.idr

```diff
@@ -60,129 +65,243 @@ genName =
     put Imps (record { nextName = i + 1 } s)
     pure $ MN "imp_gen" i
 
-mutual
+-- Processing an Idris2 expression results in
+-- an `ImperativeStatement` (since this is a monoid,
+-- this might actually be many statements) of utility
+-- functions and intermediary vars and results, and
+-- a final imperative expression, calculating a result
+-- from the related statement(s). These
+-- are then converted to JS code.
+--
+-- The boolean flag indicates, whether the two parts
+-- should be returned as a pair for further
+-- processing, or should be converted
+-- to a single statement (in which case, the final
+-- expression is converted to a `return` statement).
+ImperativeResult : (toReturn : Bool) -> Type
+ImperativeResult True  = ImperativeStatement
+ImperativeResult False = (ImperativeStatement, ImperativeExp)
+
+-- when invoking `impExp`, in some cases we always
+-- generate a pair (statements plus related final
+-- expression). In such a case, this function converts
+-- the statement to the correct result type as indicated
+-- by `toReturn`.
+pairToReturn :  (toReturn : Bool)
+             -> (ImperativeStatement, ImperativeExp)
+             -> Core (ImperativeResult toReturn)
+pairToReturn False (s, e) = pure (s, e)
+pairToReturn True (s, e)  = pure $ s <+> ReturnStatement e
```

Comment:
> Nice! One thing that comes to mind reading about Kosaraju is the fact that we now have
> two algorithms computing strongly connected components on a directed graph (the other
> one is Tarjan in `Compiler.Separate`).
> 
> We may want to consider picking one of these, moving it into `Libraries/` and using it
> everywhere.

I'll have a look at this. Since we are in the process of deciding how to continue with tail-call optimization, this might be done as part of a later PR if that's OK for you.

Reply:
Oh yeah for sure, it was more of a general comment given you're establishing a TODO list.

Reply:
Made this pure (and added lots of `pure`s in the process, haha).